bigquery_etl/glam/templates/probe_counts_v1.udf.histogram.sql

```diff
@@ -0,0 +1,161 @@
+CREATE TEMP FUNCTION udf_exponential_buckets(min FLOAT64, max FLOAT64, nBuckets FLOAT64)
+RETURNS ARRAY<FLOAT64>
+LANGUAGE js
+AS
+  '''
+  let logMax = Math.log(max);
+  let current = min;
+  if (current === 0) {
+    current = 1;
+  } // If starting from 0, the second bucket should be 1 rather than 0
+  let retArray = [0, current];
+  for (let bucketIndex = 2; bucketIndex < Math.min(nBuckets, max, 10000); bucketIndex++) {
+    let logCurrent = Math.log(current);
+    let logRatio = (logMax - logCurrent) / (nBuckets - bucketIndex);
+    let logNext = logCurrent + logRatio;
+    let nextValue =  Math.round(Math.exp(logNext));
+    current = nextValue > current ? nextValue : current + 1;
+    retArray[bucketIndex] = current;
+  }
+  return retArray
+''';
+
+CREATE TEMP FUNCTION udf_linear_buckets(min FLOAT64, max FLOAT64, nBuckets FLOAT64)
+RETURNS ARRAY<FLOAT64>
+LANGUAGE js
+AS
+  '''
+  let result = [0];
+  for (let i = 1; i < Math.min(nBuckets, max, 10000); i++) {
+    let linearRange = (min * (nBuckets - 1 - i) + max * (i - 1)) / (nBuckets - 2);
+    result.push(Math.round(linearRange));
+  }
+  return result;
+''';
+
+CREATE TEMP FUNCTION udf_functional_buckets(
+  log_base INT64,
+  buckets_per_magnitude INT64,
+  range_max INT64
+)
+RETURNS ARRAY<FLOAT64>
+LANGUAGE js
+AS
+  '''
+  function sample_to_bucket_index(sample) {
+    // Get the index of the sample
+    // https://github.com/mozilla/glean/blob/master/glean-core/src/histogram/functional.rs
+    let exponent = Math.pow(log_base, 1.0/buckets_per_magnitude);
+    return Math.ceil(Math.log(sample + 1) / Math.log(exponent));
+  }
+
+  let buckets = new Set([0]);
+  for (let index = 0; index < sample_to_bucket_index(range_max); index++) {
+    // Avoid re-using the exponent due to floating point issues when carrying
+    // the `pow` operation e.g. `let exponent = ...; Math.pow(exponent, index)`.
+    let bucket = Math.floor(Math.pow(log_base, index/buckets_per_magnitude));
+    buckets.add(bucket);
+  }
+
+  return [...buckets]
+''';
+
+CREATE TEMP FUNCTION udf_to_string_arr(buckets ARRAY<INT64>)
+RETURNS ARRAY<STRING> AS (
+  (SELECT ARRAY_AGG(CAST(bucket AS STRING)) FROM UNNEST(buckets) AS bucket)
+);
+
+CREATE TEMP FUNCTION udf_get_buckets(
+  metric_type STRING,
+  range_min INT64,
+  range_max INT64,
+  bucket_count INT64,
+  histogram_type STRING
+)
+RETURNS ARRAY<INT64> AS (
+  (
+    WITH buckets AS (
+      SELECT
+        CASE
+        WHEN
+          metric_type = 'timing_distribution'
+        THEN
+          udf_functional_buckets(2, 8, range_max)
+        WHEN
+          metric_type = 'memory_distribution'
+        THEN
+          udf_functional_buckets(2, 16, range_max)
```

Comment:
I'm just wondering if these values that are hardcoded for `log_base` and `buckets_per_magnitude` - 2 and 16 (and 2 and 8 above) are documented somewhere for glean? Might be useful to link to documentation.

Reply:
Yes, these are in the relevant pages for the distributions. I'll comment them into the code.